### DISCUSSION TITLE
How to share Jupyter Notebooks (and other tips and tricks for notebooks)

### DISCUSSION THREAD

**Discussion by shaunwbell** [NOAA-PMEL/WRC-TUG#9] [General]
As useful as Jupyter Notebooks are for combining discussion (via markdown) and code (via any of the potential integrated kernels... python, r, julia  and many others) with output (images, results), it is not always transparent what the best way to share these with others is.  

There are many options but two readily available ones are:
- Share the `.ipynb` This allows the collaborator to edit and run the original resource but needs a Jupyter instance on both machines.  [Visual Studio Code](https://code.visualstudio.com), [Google Colab](https://colab.research.google.com) and other local or cloud based IDE's may have built in Jupyter engines to render, but will still need all integrated packages to run the notebook as it was designed on the original machine.
- include the `.ipnyb` in a github repo will allow github or NBviewer to render it into html for others to view (and if certain packages are used, interact) but not edit.

If anyone has any other suggestions or comments - especially experiences of collaborative notebooks (stay tuned for AWS-centric solutions here), feel free to discuss.

**Comment by darynwhite** [comment]
I'm wondering if anyone has used [Jupyter Notebook Convert](https://github.com/jupyter/nbconvert) and has succeeded?

I've looked at using it to create a rendered, readable version of a working notebook once the work within is complete, primarily for data QC.

**Reply by shaunwbell** [reply]
I haven't used it explicitly in quite some time but it is the backend for exporting notebooks in a variety of formats when using the web interface (or Jupyter Labs). 

It generates nice html content but I think the pdf formatting is lacking. This is probably due to the fact that I don't make my notebooks with 8.5x11 paper in mind so things get compressed or blown up, etc. 

Do you have a specific output format in mind? I may have more opinions if you do ;) (well we know I always have opinions).

**Reply by shaunwbell** [reply]
Plus you might play with [https://github.com/jupyter/nbviewer](https://github.com/jupyter/nbviewer) which is a webservice version for dynamic rendering of GitHub hosted books.

**Reply by darynwhite** [reply]
Yeah, most use cases for a notebook are not for printing. This is less for actually printing it and more for non-programming users to view the reasoning of why certain flags were chosen. 

TL;DR
I need to enable people without IDEs, or Jupyter installed to understand why data was flagged a specific way, even years from when the decision was made.

**Reply by shaunwbell** [reply]
Then I think rendering it to html may be your best bet - should be readable years down the road and reasonably resemble your formatting when you developed it without the physical page size restrictions 

It's also done with jinja templates so you may be able to roll your own output look and feel.

**Reply by darynwhite** [reply]
Hmm... I had not considered rendering an HTML document over a PDF. That should function pretty well with plots, which is the critical difference between simply going with an ASCII text document as we do now.

And this is why these discussion boards can prove extremely useful!

**Comment by derekcoffman** [comment]
Not so much sharing but, rather, collaborative. I played around with this a
bit before they incorporated it into the server:
https://jupyterlab.readthedocs.io/en/stable/user/rtc.html. Not sure if it's
a jupyterlab only thing but kind of cool.

On Wed, Nov 30, 2022 at 4:27 PM shaunwbell ***@***.***> wrote:

> Plus you might play with https://github.com/jupyter/nbviewer which is a
> webservice version for dynamic rendering of GitHub hosted books.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/NOAA-PMEL/WRC-TUG/discussions/9#discussioncomment-4279038>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AFFQ3IJWRTMUKCVJCVCAL6DWK7WHFANCNFSM5L4Q3C7Q>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>


-- 
_________________________________________________________
Derek Coffman                   ***@***.***
NOAA/PMEL                       phone:  206-526-6574
7600 Sand Point Way NE
Seattle, WA  98115             fax:       206-526-6744

http://saga.pmel.noaa.gov
_________________________________________________________

**Reply by shaunwbell** [reply]
So thats like a true "Google Docs" variation of collaboration? Thats kinda cool.  Have you tested it on campus?

**Comment by derekcoffman** [comment]
No...just locally on my laptop. It wasn't perfect but even at that point it
worked pretty well. Would be interesting to see in a jupyterhub setup (if
possible).

On Thu, Dec 1, 2022 at 1:49 PM shaunwbell ***@***.***> wrote:

> So thats like a true "Google Docs" variation of collaboration? Thats kinda
> cool. Have you tested it on campus?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/NOAA-PMEL/WRC-TUG/discussions/9#discussioncomment-4287246>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AFFQ3IKV57Q2OA6KSP2NNRLWLEMO5ANCNFSM5L4Q3C7Q>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>


-- 
_________________________________________________________
Derek Coffman                   ***@***.***
NOAA/PMEL                       phone:  206-526-6574
7600 Sand Point Way NE
Seattle, WA  98115             fax:       206-526-6744

http://saga.pmel.noaa.gov
_________________________________________________________
